### DISCUSSION TITLE
[Feature Request] Wyze Thermostat Implementation information

### DISCUSSION THREAD

**Discussion by AndrewLemons** [SecKatie/ha-wyzeapi#103] [General]
The Wyze Thermostat has just been released. I currently have one, and am looking for it to be supported by Home Assistant.

I'm not sure whether control of the thermostat is possible. However, I do know for sure that some of the sensors can be read from. Here are the properties mapped to their values (the ones I currently know about, at least)

| Property | Value |
|-|-|
| `P2602` | State `1 = Home` `2 = Away` `3 = Sleep` | 
| `P2603` | Mode `0 = Off` `1 = Heat` `2 = Cool` `3 = Auto` |
| `P2604` | Cool to |
| `P2605` | Heat to |
| `P2606` | Current temperature |
| `P2607` | Units `0 = C` `1 = F` |

Here is an example response
``` JSON
[
  { "pid": "P2601", "value": "0", "ts": 1608420519245 },
  { "pid": "P2602", "value": "1", "ts": 1608430107789 },
  { "pid": "P2603", "value": "3", "ts": 1608430244679 },
  { "pid": "P2604", "value": "73", "ts": 1608430244619 },
  { "pid": "P2605", "value": "68", "ts": 1608430244632 },
  { "pid": "P2606", "value": "68.3", "ts": 1608424879880 },
  { "pid": "P2607", "value": "1", "ts": 1608428378629 },
  { "pid": "P5", "value": "1", "ts": 1608418922630 }
]
```
The information above was retrieved using the Node JS package for the API. I don't know if it alters the raw response, so take it as you will.

As a small side note, the MAC address of the device is oddly long. I was under the impression that all MACs were of a common length, but apparently not. Example `CO_EA1_xxxxxxxxxxxxxxxxxxxxxxxxxxx`.

I will gladly study the code base to see if I can implement this myself. But, if anyone else (who already knows what they are doing) can add support, that would be great.

-Andrew

**Comment by jfarmer08** [comment]
You need to get the full output from get_object_list before someone could add it. 

You could also share the device and someone could open there app to get the result of get object.

**Comment by AndrewLemons** [comment]
Here ya' go.

``` json
{
"mac":"CO_EA1_xxxxxxxxxxxxxxxxxxxxxxxx",
"first_activation_ts":1608418333000,
"first_binding_ts":1608418333000,
"enr":"",
"nickname":"Upstairs Thermostat",
"timezone_name":"UTC",
"product_model":"CO_EA1",
"product_model_logo_url":"",
"product_type":"Thermostat",
"hardware_ver":"0.0.0.0",
"firmware_ver":"1.0.0",
"user_role":1,
"binding_user_nickname":"xxx",
"conn_state":1,
"conn_state_ts":1608522611426,
"push_switch":1,
"device_params":{},
"is_in_auto":0,
"event_master_switch":1,
"parent_device_mac":"",
"parent_device_enr":"",
"binding_ts":1608418332000,
"timezone_gmt_offset":0
}
```

**Comment by SecKatie** [comment]
Awesome I will work on this today! Thanks @AndrewLemons

**Reply by mlhfb** [reply]
My thermostat is still in the box because I was disappointed it didn't have any HA integration.  I'm not looking to control it yet with HA, but I would like HA to know when the thermostat calls for service so I can turn on a ceiling fan.  This is the only project I've seen to date working on the thermostat integration!  Thank you so much!

**Reply by SecKatie** [reply]
Hey @AndrewLemons @mlhfb sorry for no updates over the last few weeks. Some family members got COVID and my mind has been preoccupied. I want to start getting back in front of the keyboard soon!

**Reply by mlhfb** [reply]
@JoshuaMulliken no worries, family first.

**Reply by yfaber** [reply]
@JoshuaMulliken i hope your fam is doing better.

**Reply by SecKatie** [reply]
Thank you everyone for your concern. They have made a full recovery! ❤️

**Comment by jfarmer08** [comment]
@JoshuaMulliken when you have time to help setup my environment I can start helping again.

**Reply by SecKatie** [reply]
@jfarmer08 Sweet. We can set aside some time this weekend if you are down for it

**Reply by SecKatie** [reply]
I am concerned that the thermostat might use a different SDK, much like many of the newer Wyze devices... do you have one @jfarmer08?

**Reply by jfarmer08** [reply]
I don't have a thermostat, but I might have someone that does that will share it so I can make the api calls. I have the vacuum that I would also like to add. Not sure but I think it's using different calls.

**Reply by SecKatie** [reply]
If you can intercept and play with the vacuum calls, maybe we can get it working.

It is super annoying, though, that the devices do not use the same control mechanism as their earlier devices.

**Reply by BoredMind** [reply]
I have every product Wyze offers, I'm willing to help in any way I can.  I've got Windows, Mac, and Linux environments here ready to go.

**Reply by SecKatie** [reply]
Sweet! Can you use Burp Suite? I usually record the calls to the server and modify them to see the minimal amount of information needed to make the thing work.

I am happy to help any owners of the newer devices with this. We can set up some zoom calls or something.

**Reply by jfarmer08** [reply]
@JoshuaMulliken yeah very new product is just a rebrand and they are just making new calls instead of using current methods. This will come back to hunt them. I will start pulling data for the vacuum. If you do any zoom call and I am free I would be glad to join and help.

**Reply by LucasPMagno** [reply]
I'd be glad to help as well. I have mitmproxy setup to intercept calls from the wyze vacuum.

**Reply by BoredMind** [reply]
I can capture using mitmproxy or Burp.  Probably best to do a zoom so you can see the data coming through and what you need or not.

**Reply by beardedoffender** [reply]
I have most wyze products, including the thermostat and vacuum with the video doorbell on preorder. I am more than willing to assist. I've setup burpsuite but am not familiar with it. I'm currently going over the documentation. I am always available for a hangout or zoom call if anyone wants to work on this

**Reply by cuatexoft** [reply]
How do you got the codes @AndrewLemons ? Can you get it for Outdoor plug ? Thanks

**Reply by AndrewLemons** [reply]
I used the Node JS version of the wyze api to get the properties of my thermostat. Figuring out what they meant was just trial and error. I would get them for the outdoor plugs but I do not currently have one. I would like to imagine that they would be similar to the regular smart plug though.

**Reply by SecKatie** [reply]
Apparently, the outdoor plug is easy so I went ahead and added the code that should make it supported there is a pull request. I am just waiting on review before I merge it

**Reply by jfarmer08** [reply]
I will make the changes on my system and make sure everything works.

**Reply by cuatexoft** [reply]
Any update? Can you share the codes for outdoor plug ?  Thanks

**Reply by SecKatie** [reply]
@cuatexoft there is a beta release. Please install that and inform me if it is working.

**Reply by jscherry** [reply]
I tested it works perfectly

**Reply by cuatexoft** [reply]
And any update about the thermostat?

**Reply by cuatexoft** [reply]
Hi Joshua, do you have an update on this?  Im wonder if you could oublish a video of how this will work in HA.  thank you for all your work.

**Reply by mlhfb** [reply]
> Sweet! Can you use Burp Suite? I usually record the calls to the server and modify them to see the minimal amount of information needed to make the thing work.
> 
> I am happy to help any owners of the newer devices with this. We can set up some zoom calls or something.

Is there a good tutorial for getting a Burp MITM going?  Would I just proxy enable my phone with the app on it?  I spent some time figuring out how to capture just the Wyze Thermostat traffic with Wireshark through my Unifi AP, but I cand make any sense of the hex or ascii data.  I'm a green horn to say the least but I can trial and error things out.  I mean I knew very little about SSH Key authentication until yesterday.  Anyway I'm willing to set up the tools and capture data, test out code.  I have dabbled in writing Arduino C++, I've done less python, but with examples and Stack Overflow I can get the job done. 10x slower, and 100x messier than a trained dev I'm sure.   Let me know if I can help.

**Reply by p1ironman** [reply]
Has this implementation become stale? I have the thermostat and would live to have it integrated into Home Assistant.

**Comment by p1ironman** [comment]
I have the thermostat, not sure what i can do to help get this complete. Please advise.

**Comment by Z1mDMan** [comment]
I have one of the thermostats and am willing to help if I can. Mainly macOS here.

**Comment by mlhfb** [comment]
@JoshuaMulliken is it possible to just integrate the status of the thermostat?  Such as set temp, if calling for heat, if calling for A/C, is system fan on, etc.  I'm willing to help and while I may not be dev, I know enough to be dangerous ;)  Thanks for your contribution to the community!

**Comment by pabo** [comment]
> As a small side note, the MAC address of the device is oddly long. I was under the impression that all MACs were of a common length, but apparently not. Example CO_EA1_xxxxxxxxxxxxxxxxxxxxxxxxxxx.

This appears to be the model number (`CO_EA1`) followed by the serial number (`313046........`). I don't see a straightforward correlation between the serial number and the MAC address, and I don't see the MAC address reported by the API anywhere. But the device does have a MAC address with the proper vendor prefix `2C:AA:8E...`, which is reported by the Wyze app after registering the device, and shown in network traffic if you sniff it.

Not sure why they've labeled this identifier `mac` in the API response

**Comment by pabo** [comment]
I got my thermostat today and installed it, and registered it in the app. I set up a SSL MITM via Charles to sniff out what was happening between my phone and the wyze servers.

Loading the thermostat in the app hits some URLs at `https://wyze-earth-service.wyzecam.com` like this:
```
path: /plugin/earth/device_info?device_id=CO_EA1_3130XXXXXXXXXXXXXXXXXXXX&keys=plugin_version&nonce=XXXXXXXXXXXX
```

```
path: /plugin/earth/get_iot_prop?did=CO_EA1_3130XXXXXXXXXXXXXXXXXXXX&keys=iot_state%2Capp_version&nonce=XXXXXXXXXXXX
```

```
path: /plugin/earth/device_info?device_id=CO_EA1_3130XXXXXXXXXXXXXXXXXXXX&keys=setup_personalization_state%2Csetup_test_state%2Ctime_zonesetup_wires_selectedsetup_wires_warning1setup_wires_warning2setup_wires_warning3%2Cterminal_type&nonce=XXXXXXXXXXXX
```

but this one seems to be the money maker for getting the current state to show on the phone app:

```
path: /plugin/earth/get_iot_prop?did=CO_EA1_3130XXXXXXXXXXXXXXXXXXXX&keys=iot_state%2Cdevice_state%2Ctemperature%2Chumidity%2Ctime2temp_val%2Cprotect_time%2Ctemp_unit%2Cmode_sys%2Ccurrent_scenario%2Cconfig_scenario%2Cheat_sp%2Ccool_sp%2Cw_city_id%2Cw_lat%2Cw_lon%2Cstage_io_test%2Csetup_state%2Cfan_mode%2Cwiring_logic_id%2Cworking_state%2Cemheat%2Casw_hold%2Cdev_hold%2Cdev_holdtime%2Csave_comfort_balance%2Ctrigger_off_val&nonce=XXXXXXXXXXXX

response: 
{
	"code": 1,
	"current": 1,
	"data": {
		"did": "CO_EA1_3130XXXXXXXXXXXXXXXXXXXX",
		"is_online_send": false,
		"is_sub_device": false,
		"is_update_shadow": false,
		"model": "CO_EA1",
		"parent_did": null,
		"parent_model": null,
		"props": {
			"heat_sp": "68",
			"mode_sys": "off",
			"iot_state": "connected",
			"stage_io_test": "{\"G\":{\"7\":\"1\"},\"H1\":{\"1\":\"1\",\"7\":\"1\"},\"C1\":{\"3\":\"1\",\"7\":\"1\"}}",
			"setup_state": "H01",
			"w_lat": "XX.XX",
			"current_scenario": "home",
			"config_scenario": "{\"sleep\":{\"cool\":\"72\",\"heat\":\"66\"},\"home\":{\"cool\":\"76\",\"heat\":\"68\"},\"away\":{\"cool\":\"79\",\"heat\":\"56\"}}",
			"protect_time": "0",
			"emheat": "0",
			"cool_sp": "76",
			"save_comfort_balance": "4",
			"device_state": "idle",
			"time2temp_val": "0",
			"working_state": "idle",
			"wiring_logic_id": "F20200010",
			"w_city_id": "5018910",
			"fan_mode": "auto",
			"temperature": "73.6",
			"humidity": "31",
			"w_lon": "-XXX.XX",
			"dev_hold": "0",
			"temp_unit": "F"
		}
	},
	"hash": "1",
	"instance_id": "XXXXXXX",
	"message": "Success",
	"total": 1,
	"version": 1
}
```


And then if I do something like turn on the fan, I get this:

```
path: /plugin/earth/set_iot_prop_by_topic

request body:
{
	"did": "CO_EA1_3130XXXXXXXXXXXXXXXXXXXX",
	"nonce": "XXXXXXXXX",
	"props": {
		"fan_mode": "on"
	},
	"model": "CO_EA1",
	"is_sub_device": 0
}

response body:
{
	"code": 1,
	"current": 0,
	"data": null,
	"hash": "1",
	"instance_id": "XXXXXXXX",
	"message": "Success",
	"total": 0,
	"version": 1
}
```

If any of this is helpful I can do more. Just let me know what is needed.

**Comment by r14n** [comment]
I don't have any climate entities, is there some issue with my config or is there a bug?  When HA is loading, it says "Starting climate.wyzeapi, not everything will be available until it is finished." and it doesn't go away.  Any ideas?  I'm new to HA, hoping to get this working.
